### PR TITLE
dmctl: try read master addr from env

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve? 

#1720 

### What is changed and how it works?

as title

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

-None

Side effects

- None

Related changes
 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

**Comment by lance6716** [comment]
add an integration test, rest lgtm

File: tests/dmctl_command/run.sh [hunk 1]
```diff
@@ -86,6 +86,12 @@ function run() {
 	# query status with command mode
 	$PWD/bin/dmctl.test DEVEL --master-addr=:$MASTER_PORT query-status >$WORK_DIR/query-status.log
 
+	# use DM_MASTER_ADDR env
+	export DM_MASTER_ADDR="127.0.0.1:$MASTER_PORT"
```
Comment (GMHDBJD) [inline tests/dmctl_command/run.sh:90]: How about MASTER_ADDR ？

**Comment by GMHDBJD** [comment]
You can also test tiup dmctl in https://github.com/pingcap/dm/blob/0f3a7755d0e69693a99821c2981e9bbba0747bff/tests/tiup/upgrade-from-v2.sh#L80

**Comment by GMHDBJD** [comment]
You can also test tiup dmctl in https://github.com/pingcap/dm/blob/0f3a7755d0e69693a99821c2981e9bbba0747bff/tests/tiup/upgrade-from-v2.sh#L82

File: tests/dmctl_command/run.sh (hunk 1 above)
Comment (lance6716) [reply tests/dmctl_command/run.sh:90]: now I think a DM prefix is better, because     it distinguishes itself from all environment variables.

File: dm/ctl/common/config.go [hunk 2]
```diff
@@ -149,6 +150,10 @@ func (c *Config) Parse(arguments []string) (finish bool, err error) {
 		return false, errors.Errorf("'%s' is an invalid flag", c.FlagSet.Arg(0))
 	}
 
+	// try get master Addr from env "DM_MASTER_ADDR" if this flag is empty.
+	if c.MasterAddr == "" {
+		c.MasterAddr = os.Getenv("DM_MASTER_ADDR")
+	}
 	if c.MasterAddr == "" {
 		return false, errors.Errorf("--master-addr not provided, use `dmtcl --help` to see help messages")
```
Comment (lance6716) [inline dm/ctl/common/config.go:158]: update this comment to mention env

File: tests/dmctl_command/run.sh (hunk 1 above)
Comment (Ehco1996) [reply tests/dmctl_command/run.sh:90]: i think add DM as namespace is better?

File: dm/ctl/common/config.go [hunk 3]
```diff
@@ -63,7 +64,7 @@ func NewConfig() *Config {
 
 	fs.BoolVar(&cfg.printVersion, "V", false, "Prints version and exit.")
 	fs.StringVar(&cfg.ConfigFile, "config", "", "Path to config file.")
-	fs.StringVar(&cfg.MasterAddr, "master-addr", "", "Master API server address, this parameter is required when interacting with the dm-master")
+	fs.StringVar(&cfg.MasterAddr, "master-addr", "", "Master API server address, this parameter is required when interacting with the dm-master, you can use environment var 'DM_MASTER_ADDR' to pass value to this flag")
```
Comment (lance6716) [inline dm/ctl/common/config.go:67]: don't change flag comment because env and command line parameter are not the same thing 🤔 you only need to change the error message

**Review by lance6716 — CHANGES_REQUESTED** [review]

File: dm/ctl/common/config.go [hunk 4]
```diff
@@ -63,7 +64,7 @@ func NewConfig() *Config {
 
 	fs.BoolVar(&cfg.printVersion, "V", false, "Prints version and exit.")
 	fs.StringVar(&cfg.ConfigFile, "config", "", "Path to config file.")
-	fs.StringVar(&cfg.MasterAddr, "master-addr", "", "Master API server address, this parameter is required when interacting with the dm-master")
```
Comment (lance6716) [inline dm/ctl/common/config.go:66]: maybe revert this change

File: dm/ctl/common/config.go [hunk 5]
```diff
@@ -149,8 +150,12 @@ func (c *Config) Parse(arguments []string) (finish bool, err error) {
 		return false, errors.Errorf("'%s' is an invalid flag", c.FlagSet.Arg(0))
 	}
 
+	// try get master Addr from env "DM_MASTER_ADDR" if this flag is empty.
 	if c.MasterAddr == "" {
-		return false, errors.Errorf("--master-addr not provided, use `dmtcl --help` to see help messages")
+		c.MasterAddr = os.Getenv("DM_MASTER_ADDR")
+	}
+	if c.MasterAddr == "" {
+		return false, errors.Errorf("--master-addr not provided, this parameter is required when interacting with the dm-master, you can use environment var 'DM_MASTER_ADDR' to pass value to this flag, use `dmtcl --help` to see more help messages")
```
Comment (lance6716) [inline dm/ctl/common/config.go:158]:
```suggestion
		return false, errors.Errorf("--master-addr not provided, this parameter is required when interacting with the dm-master, you can also use environment variable 'DM_MASTER_ADDR' to specify the value. Use `dmtcl --help` to see more help messages")
```

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: tests/tiup/upgrade-from-v2.sh [hunk 6]
```diff
@@ -79,7 +79,8 @@ function migrate_in_v2 {
 	echo "check locks"
 	run_dmctl_with_retry $CUR_VER "show-ddl-locks" "no DDL lock exists" 1
 
-	tiup dmctl:$CUR_VER --master-addr=master1:8261 stop-task $TASK_NAME
+	export DM_MASTER_ADDR="master1:8261"
+	tiup --binpath="./bin/dmctl" dmctl:$CUR_VER stop-task $TASK_NAME
```
Comment (GMHDBJD) [inline tests/tiup/upgrade-from-v2.sh:83]: Can we pass the test if remove binpath? We patch tiup with locally build version at L54.

**Review by GMHDBJD — COMMENTED** [review]
rest LGTM

File: tests/tiup/upgrade-from-v2.sh (hunk 6 above)
Comment (GMHDBJD) [reply tests/tiup/upgrade-from-v2.sh:83]: Seems fail in above test. 🤔

Comment (Ehco1996) [reply tests/tiup/upgrade-from-v2.sh]: seems tiup still use  nightly version dmctl after update, how can we use the new one？
`Starting component `dmctl`: /root/.tiup/components/dmctl/v5.0.0-nightly-20210524/dmctl/dmctl stop-task`

Comment (GMHDBJD) [reply tests/tiup/upgrade-from-v2.sh]: I will have a look later.

Comment (GMHDBJD) [reply tests/tiup/upgrade-from-v2.sh]: Change this [CUR_VER](https://github.com/pingcap/dm/blob/0f3a7755d0e69693a99821c2981e9bbba0747bff/tests/tiup/upgrade-from-v2.sh#L49) to PRE_VER. 🤣

File: tests/tiup/upgrade-from-v2.sh [hunk 7]
```diff
@@ -80,7 +80,7 @@ function migrate_in_v2 {
 	run_dmctl_with_retry $CUR_VER "show-ddl-locks" "no DDL lock exists" 1
 
 	export DM_MASTER_ADDR="master1:8261"
-	tiup dmctl:$CUR_VER stop-task $TASK_NAME
+	tiup dmctl:$PRE_VER stop-task $TASK_NAME
```
[the remark this answers is not in the document]
Comment (Ehco1996) [reply tests/tiup/upgrade-from-v2.sh]: ci still failed, seems `PRE_VER` use dmctl v2.0.3


Starting component `dmctl`: /root/.tiup/components/dmctl/v2.0.3/dmctl/dmctl pause-task upgrade_via_tiup

[the remark this answers is not in the document]
Comment (Ehco1996) [reply tests/tiup/upgrade-from-v2.sh]: also try $CUR_VER still not work

$CUR_VER=version v5.0.0-nightly-20210524 
$PRE_VER=version v2.0.3

need help Gangster

**Comment by GMHDBJD** [comment]
I mean use PRE_VER at L49, CUR_VER at L83, and test DM_MASTER_ADDR at L83.
PRE_VER is previous version, CUR_VER is nightly.
If we use CUR_VER(nightly) at L49(before we patch nightly with local build binary files), tiup will download current nightly version(master version), after we patch new nightly version, it will not upgrade and still use previous nightly version.

File: tests/tiup/upgrade-from-v2.sh [hunk 8]
```diff
@@ -46,18 +46,20 @@ function migrate_in_previous_v2() {
...
 
 function upgrade_to_current_v2() {
 	if [[ "$CUR_VER" == "nightly" && "$ref" == "refs/pull"* ]]; then
 		patch_nightly_with_tiup_mirror
 	fi
-	tiup update dmctl:$CUR_VER
 	tiup dm upgrade --yes $CLUSTER_NAME $CUR_VER
+	# uninstall previous dmctl, otherwise dmctl:nightly still use PRE_VER.
+	# FIXME: It may be a bug in tiup mirror.
+	tiup uninstall dmctl --all
```
Comment (GMHDBJD) [inline tests/tiup/upgrade-from-v2.sh:59]: To be investigated. 😢 PTAL now.

Comment (Ehco1996) [reply tests/tiup/upgrade-from-v2.sh:59]: thanks so so much

**Review by GMHDBJD — APPROVED** [review]
/lgtm

**Review by lonng — COMMENTED** [review]
/lgtm
